rust/otap-dataflow/benchmarks/benches/retry_processor/main.rs

```diff
@@ -0,0 +1,331 @@
+
+                        let _ = black_box(0usize);
+                    })
+                    .await;
+            });
+        },
+    );
+
+    group.finish();
+}
+
+/// Benchmark 2: HashMap performance with different numbers of pending messages
+fn bench_pending_message_operations(c: &mut Criterion) {
+    let rt = tokio::runtime::Builder::new_current_thread()
+        .enable_all()
+        .build()
+        .expect("failed to build Tokio runtime");
+
+    let mut group = c.benchmark_group("retry_processor_hashmap_scale");
+
+    for &pending_count in PENDING_MESSAGE_SCALES {
+        let _ = group.throughput(Throughput::Elements(pending_count as u64));
+
+        let _ = group.bench_function(BenchmarkId::new("hashmap_operations", pending_count), |b| {
+            b.to_async(&rt).iter(move || async move {
+                let local = LocalSet::new();
+
+                local
+                    .run_until(async {
+                        let config = RetryConfig {
+                            max_retries: 5,
+                            initial_retry_delay_ms: 1000,
+                            max_retry_delay_ms: 30000,
+                            backoff_multiplier: 2.0,
+                            max_pending_messages: pending_count * 2, // Ensure we don't hit limits
+                            cleanup_interval_secs: 3600, // Don't cleanup during benchmark
+                        };
+                        let mut processor = RetryProcessor::with_config(config);
+                        let (sender, _receiver) = mpsc::Channel::new(1000);
+                        let mut effect_handler =
+                            EffectHandler::new("bench_processor".into(), Sender::Local(sender));
+
+                        // Fill up the pending messages HashMap
+                        for i in 0..pending_count {
+                            let otlp_data = create_otlp_logs_data();
+                            let msg_id = i as u64; // Use simple sequential IDs for benchmarking
+                            processor
+                                .process(Message::PData(otlp_data), &mut effect_handler)
+                                .await
+                                .expect("failed to process PData in hashmap benchmark");
+
+                            // NACK to keep in pending state
+                            processor
+                                .process(
+                                    Message::Control(ControlMsg::Nack {
+                                        id: msg_id,
+                                        reason: "Keep pending".to_string(),
+                                    }),
+                                    &mut effect_handler,
+                                )
+                                .await
+                                .expect("failed to process NACK in hashmap benchmark");
+                        }
+
+                        // Now measure operations on the full HashMap
```

Comment:
nit: the fact that the `RetryProcessor` is using a `HashMap` is kind of an implementation detail. Maybe it doesn't make sense to leak it into the comments here?

Reply:
no worries, updated the comment and other mentions in strings, etc